### PR TITLE
[ADD] tests: add a script to test module uninstallation

### PR DESCRIPTION
**Pull request by d-fence**
Add a script to test uninstallation  of modules.
This script is a kind of standalone tool. It uses odoo as a library but the odoo server is not started at all.

In its standard invocation, it tries an install/uninstall/reinstall cycle for each all module found in the specified database.

By specifying '-U', it only tries to uninstall the comma separated list of modules following the argument.

That way, this tool can be used on the runbot in nightly builds

### CODE REVIEW

File: odoo/tests/test-uninstalls.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+            # order modules in topological order
+            modules = modules.browse(topological_sort({
+                module.id: module.dependencies_id.depend_id.ids
+                for module in modules
+            }))
+            module_names = {module.id: module.name for module in modules}
+
+    resuming = args.resume
+    for module_id, module_name in module_names.items():
```
Comment (odony) [inline odoo/tests/test-uninstalls.py]:
```suggestion
    # `module_names` preserves the order of `modules` in P3.6+
    for module_id, module_name in module_names.items():
```

File: odoo/tests/test-uninstalls.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Script for testing the install / uninstall / reinstall cycle of Odoo modules")
+    parser.add_argument("--database", "-d", type=str, required=True,
+        help="The database to test (note: must have only 'base' installed)")
+    parser.add_argument("--skip", "-s", type=str, nargs=1,
+        help="Comma-separated list of modules to skip (they will only be installed)")
+    parser.add_argument("--resume", "-r", type=str, nargs=1,
+        help="Skip the cycle on all modules up to the specified one")
```
Comment (odony) [inline odoo/tests/test-uninstalls.py]:
```suggestion
        help="Skip modules (only install) up to the specified one in topological order")
```
or something else that indicates the order

File: odoo/tests/test-uninstalls.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    uninstall(db_name, module_id, module_name)
+    install(db_name, module_id, module_name)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Script for testing the install / uninstall / reinstall cycle of Odoo modules")
+    parser.add_argument("--database", "-d", type=str, required=True,
+        help="The database to test (note: must have only 'base' installed)")
+    parser.add_argument("--skip", "-s", type=str, nargs=1,
```
Comment (odony) [inline odoo/tests/test-uninstalls.py]: there's probably a reason but it's not obvious to me why the `nargs=1`, which means we have to use `args.arg[0]` in both cases?

File: odoo/tests/test-uninstalls.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        help="Comma-separated list of modules to skip (they will only be installed)")
+    parser.add_argument("--resume", "-r", type=str, nargs=1,
```
Comment (odony) [inline odoo/tests/test-uninstalls.py]: Just a thought: maybe "--resume-at" would more clearly show that the specified module will be part of the cycle?
```suggestion
    parser.add_argument("--resume-at", "-r", type=str, nargs=1,
```

File: odoo/tests/test-uninstalls.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+        if skip:
+            install(args.database, module_id, module_name)
+        else:
+            cycle(args.database, module_id, module_name)
+
+
+def test_uninstall(args):
+    """ Tries to uninstall/reinstall one ore more modules"""
+    domain = [('name', 'in', args.uninstall.split(',')), ('state', '=', 'installed')]
```
Comment (odony) [inline odoo/tests/test-uninstalls.py]: should we silently ignore modules that aren't installed? depends on the use case for this mode...

File: odoo/tests/test-uninstalls.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                skip = True
+
+        # handle skip option
+        if args.skip and module_name in args.skip[0].split(','):
+            skip = True
+
+        if skip:
+            install(args.database, module_id, module_name)
+        else:
+            cycle(args.database, module_id, module_name)
```
Comment (odony) [inline odoo/tests/test-uninstalls.py]: Would you like some Parmesan? ;-)

Can we make this more straightforward, e.g. something like this?

```python
resume = args.resume and args.resume[0]
skip = args.skip and set(args.skip[0].split(','))

for module_id, module_name in module_names.items():
    # stop resuming if we've reached the target
    if module_name == resume:
        resume = None

    if resume or module_name in skip:
        install(args.database, module_id, module_name)
    else:
        cycle(args.database, module_id, module_name)
```

Or even better without `nargs=1` ?

**Comment by d-fence** [comment]
> Didn't test but it looks nice and simple enough, only have minor nitpicking comments.
> 
> I take it the intended nightly mode is the full cycle, and the other options (resume, skip, uninstall-list) are for debugging? Or did you have a specific use case for those?
> 
> PS: don't ask me why the build failed with obscure upgrade errors :p

The full cycle may take a while, I think that we will run it weekly.
The uninstall is made to test uninstall from a fully installed database, it would be run nightly.
The actual "uninstall" runbot configuration is set on the branch and the result is already available.
e.g. on the last push from RCO: http://runbot.odoo.com/runbot/build/1524004

File: odoo/tests/test-uninstalls.py (hunk 5 above)
Comment (d-fence) [reply odoo/tests/test-uninstalls.py]: The use case is for the runbot, it creates a database with everything installed and then create subbuilds that restaore this full database en tries to uninstall each module in a separate build.
So, yes, we silently ignore module that aren't installed.

File: odoo/tests/test-uninstalls.py (hunk 1 above)
Comment (rco-odoo) [reply odoo/tests/test-uninstalls.py]: I will use a plain list instead.  The dict is not necessary anymore.

**Comment by d-fence** [comment]
@Elkasitu are the changes ok for you ?

**Comment by d-fence** [comment]
> LGTM but maybe we should use a command-line option for the blacklist / ignore list instead of having them hardcoded so that it's easily settable on a runbot configuration? Like how the runbot implements its blacklist / ignore list currently (i think?)

In the current runbot config, a full install is done and then the produced database is used in sub-builds. Each one tries to remove a single module from the database with the `-U` argument.
So, the blacklist could be made in the runbot step that choose the modules to uninstall.

**Comment by Elkasitu** [comment]
> > LGTM but maybe we should use a command-line option for the blacklist / ignore list instead of having them hardcoded so that it's easily settable on a runbot configuration? Like how the runbot implements its blacklist / ignore list currently (i think?)
> 
> In the current runbot config, a full install is done and then the produced database is used in sub-builds. Each one tries to remove a single module from the database with the `-U` argument.
> So, the blacklist could be made in the runbot step that choose the modules to uninstall.

Fair enough for `test_uninstall` but the same doesn't apply to `test_full` (which for me is the primary case) right?

**Comment by d-fence** [comment]
@rco-odoo can we rplus this one ?

**Pull request closed without merging**